### ISSUE TITLE
Minimal complete example of file transfer

### ISSUE THREAD

**Issue by gwgundersen** [globus/globus-sdk-python#304]
I would like to transfer some files programmatically between two remote clusters. I cannot find a simple example of how to do this.

To explain my confusion: I have used APIs before, but I am used to being given a secret key or credential through a web interface and then doing something like:

```
from service import Client
secret_key = os.environ['SERVICE_SECRET_KEY']
client = Client(secret_key)
client.make_api_call()
```

The Globus example [here](https://globus-sdk-python.readthedocs.io/en/stable/tutorial/#) is more complicated. For example, it looks like I generate a token using Python, but I have no way of programmatically getting the token because I cannot open a web browser on my cluster. When I try this locally so I can open the link using a web browser, I get a URI mismatch error.

When I try to use the `ClientCredentialsAuthorizer` class, my transfer initialization fails because of a `'Token is not active'` error.

It seems like my issue is generating and storing a token programmatically. Sorry this is broad. Thanks.

**Comment by sirosen** [comment]
> I am used to being given a secret key or credential through a web interface and then doing something like:
> [code snippet]

I agree that this mode of usage would be much better for developers and scripting use cases. We unfortunately don't have that capability at present, so, working with what is currently available...

We need to document the shortest path(s) to getting and using credentials for automation.
So I think there's a significant documentation improvement to do ( e.g. #256 if I can ever get the time), but I can also try to answer your issue to get you up an running here. Maybe if this answer is good enough it will be a basis for a section in the docs.

The [SDK tutorial](https://globus-sdk-python.readthedocs.io/en/stable/tutorial/) _is_ more complicated than fetching an API key, Personal Access Token, or other simple, static credential. However, it can be manipulated into producing a similar end result.

The basic flow you want will probably be to create an app, login to it on your laptop or other device with a functioning browser, and then to get the credentials resulting from login + the app's registration info to your cluster or other remote environment.
- You need to use refresh tokens for long term credentials, which can be fetched with a simple option to the login flow, as detailed in [one of the advanced sections of the tutorial](https://globus-sdk-python.readthedocs.io/en/stable/tutorial/#advanced-2-refresh-tokens-never-login-again)
- This can be done by following the SDK Tutorial and then adding the refresh token flag
- Your browser doesn't need to be open on the machine which will receive your refresh token(s), here's how this can work:
  - remote machine without graphical session (e.g. SSH) has client ID info from app registration
  - remote machine generates a login link, per Tutorial content
  - you copy that login link to your browser, on the local machine
  - you do a login flow in your browser, all local, and follow it until it takes you to https://auth.globus.org/v2/web/auth-code
  - you take the resulting Authorization Code and paste it to the remote machine
  - remote machine exchanges the Authorization Code for Refresh Tokens
- Pass the refresh token(s) to [RefreshTokenAuthorizer](https://globus-sdk-python.readthedocs.io/en/stable/authorization/#globus_sdk.RefreshTokenAuthorizer) objects -- this encodes your Authorization
- Pass said `RefreshTokenAuthorizer` to the client class you want to use, presumably `TransferClient`

You can see an example of this very flow by installing the Globus CLI (`pip install --user globus-cli`) on a remote machine and running `globus login`. Trying this out to see how it works would be a good step in getting familiar with what the flow looks like.

There's another way of handling credentials which might interest you, which is to use Client Credentials, per [this example](https://globus-sdk-python.readthedocs.io/en/stable/examples/client_credentials/#using-clientcredentialsauthorizer).
This flow is simpler than the interactive login flow -- it's just exchanging a client ID + client secret for tokens -- but it has the disadvantage that it gives you tokens for the "user" `<client ID>@clients.auth.globus.org`, so you may need to grant that user additional permissions (vs using your own user and its associated permissions).

**Comment by gwgundersen** [comment]
Thanks for this. After following your instructions, I was able to authenticate properly.

As additional documentation for anyone who stumbles upon this issue, here is a small module I wrote that combines various snippets from the documentation in one place. It expects your app client ID to be saved as an environment variable and the refresh token to live in a file called `/path/to/transfer_rt.txt`. These can both be handled differently depending on your security situation:

```python
# ------------------------------------------------------------------------------
# globus_auth.py
# ------------------------------------------------------------------------------

import globus_sdk
from   globus_sdk.exc import AuthAPIError
import os

# ------------------------------------------------------------------------------

def get_transfer_client():
    RT_FNAME  = '/path/to/transfer_rt.txt'
    CLIENT_ID = os.environ.get('GLOBUS_CLIENT_ID')
    client    = globus_sdk.NativeAppAuthClient(CLIENT_ID)
    client.oauth2_start_flow(refresh_tokens=True)

    try:
        with open(RT_FNAME, 'r') as f:
            refresh_token = f.read()
            rt_authorizer = globus_sdk.RefreshTokenAuthorizer(refresh_token,
                                                              client)
    except (IOError, AuthAPIError) as e:
        refresh_token = authenticate_using_browser(client)
        with open(RT_FNAME, 'w+') as f:
            f.write(refresh_token)
        rt_authorizer = globus_sdk.RefreshTokenAuthorizer(refresh_token, client)
    transfer_client = globus_sdk.TransferClient(authorizer=rt_authorizer)
    return transfer_client

# ------------------------------------------------------------------------------

def authenticate_using_browser(client):
    url = client.oauth2_get_authorize_url()
    print('Please go to this URL and login: %s' % url)
    get_input      = getattr(__builtins__, 'raw_input', input)
    auth_code      = get_input('Please enter the code here: ').strip()
    token_response = client.oauth2_exchange_code_for_tokens(auth_code)
    refresh_token  = token_response.get('refresh_token')
    return refresh_token
```

Allowing me to do the following in any script:

```python
import globus_auth
transfer_client = globus_auth.get_transfer_client()
```

Thanks again.

**Comment by gwgundersen** [comment]
Apparently, that was not enough to complete a file transfer. Using [documentation for the transfer client](https://globus-sdk-python.readthedocs.io/en/stable/clients/transfer/), I've tried:

```python
import globus_sdk
import globus_auth

tc    = globus_auth.get_transfer_client()
tdata = globus_sdk.TransferData(tc, SOURCE_EP_ID, TARGET_EP_ID)
```

I get the following error:

```
globus_sdk.exc.TransferAPIError: (401, 'AuthenticationFailed', 'Token is not active', '3kxIK5p24')
```

I've also tried to follow the documentation for authenticating the endpoint using my web browser, but the web page for my endpoint just loads fine, and I'm not sure what to do next.

Maybe what I'm trying to do is not a good use-case for Globus?
